staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go

```diff
@@ -619,6 +621,248 @@ func RunTestPreconditionalDeleteWithOnlySuggestionPass(ctx context.Context, t *t
+	}{
+		{
+			name:  "RV=Empty",
+			match: "",
+		},
+		{
+			name:  "RV=NotOlderThan",
+			match: metav1.ResourceVersionMatchNotOlderThan,
+		},
+		{
+			name:  "RV=MatchExact",
+			match: metav1.ResourceVersionMatchExact,
+		},
+	}
+
+	for _, c := range cases {
+		b.Run(c.name, func(b *testing.B) {
+			runBenchmarkStoreList(ctx, b, store, 0, maxRevision, c.match, false, nodeNames)
+		})
+	}
+	b.Run("Paginate", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreList(ctx, b, store, paginateLimit, maxRevision, c.match, false, nodeNames)
+			})
+		}
+	})
+	b.Run("NodeIndexed", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreList(ctx, b, store, 0, maxRevision, c.match, true, nodeNames)
+			})
+		}
+		b.Run("Paginate", func(b *testing.B) {
+			for _, c := range cases {
+				b.Run(c.name, func(b *testing.B) {
+					runBenchmarkStoreList(ctx, b, store, paginateLimit, maxRevision, c.match, true, nodeNames)
+				})
+			}
+		})
+	})
+	b.Run("Namespace", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreListNamespace(ctx, b, store, maxRevision, c.match, namespacedNames)
+			})
+		}
+	})
+}
+
+func runBenchmarkStoreListNamespace(ctx context.Context, b *testing.B, store storage.Interface, maxRV int, match metav1.ResourceVersionMatch, namespaceNames []string) {
+	wg := sync.WaitGroup{}
+	objectCount := atomic.Uint64{}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		resourceVersion := ""
+		switch match {
+		case metav1.ResourceVersionMatchExact:
+			resourceVersion = fmt.Sprintf("%d", maxRV-99+i%100)
+		case metav1.ResourceVersionMatchNotOlderThan:
+			resourceVersion = "0"
+		}
+		go func(resourceVersion string) {
+			defer wg.Done()
+			for j := 0; j < len(namespaceNames); j++ {
+				listOut := &example.PodList{}
+				err := store.GetList(ctx, "/pods/"+namespaceNames[j], storage.ListOptions{
+					Recursive:            true,
+					ResourceVersion:      resourceVersion,
+					ResourceVersionMatch: match,
+					Predicate:            storage.Everything,
+				}, listOut)
+				if err != nil {
+					panic(fmt.Sprintf("Unexpected error %s", err))
+				}
+				objectCount.Add(uint64(len(listOut.Items)))
+			}
+		}(resourceVersion)
+	}
+	wg.Wait()
+	b.ReportMetric(float64(objectCount.Load())/float64(b.N), "objects/op")
+}
+
+func runBenchmarkStoreList(ctx context.Context, b *testing.B, store storage.Interface, limit int64, maxRV int, match metav1.ResourceVersionMatch, perNode bool, nodeNames []string) {
+	wg := sync.WaitGroup{}
+	objectCount := atomic.Uint64{}
+	pageCount := atomic.Uint64{}
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		resourceVersion := ""
+		switch match {
+		case metav1.ResourceVersionMatchExact:
+			resourceVersion = fmt.Sprintf("%d", maxRV-99+i%100)
+		case metav1.ResourceVersionMatchNotOlderThan:
+			resourceVersion = "0"
```

Comment:
This may potentially lead to different results - in pathological case we may return from empty watchcache.
I actually suggest using the exact same RV as above.

Reply:
Done